### PR TITLE
support OnCompactionBegin

### LINKED ISSUE #4288: Cannot intercept the Compaction-Begin event
In my project I use the rocksdb::EventListener::GetCompactionEventListener function as a callback for the Compaction-Begin event. I'm on RocksDB 5.8.

But, that callback has been removed from RocksDB 5.13.1 with commit https://github.com/facebook/rocksdb/commit/1209b6db5c5ea3e26d72a12e94b3d9cf6c93f6c5 .

So, from v5.13.1 we have this in rocksdb::EventListener:
- OnFlushBegin
- OnFlushCompleted
- OnCompactionCompleted

Note that my specific use case is to use some thread-local variables to track things that happen in RocksDB Background threads (Flush and Compaction). That is why the OnCompactionCompleted is not enough.

### Expected behavior
To have an OnCompactionBegin callback or any other way for intercepting that event.

### Actual behavior
We cannot intercept the Compaction-Begin event.

### PR DESCRIPTION
**Pull request by solicomo**
fix #4288 

Add `OnCompactionBegin` support to `rocksdb::EventListener`.

Currently, we only have these three callbacks:

- OnFlushBegin
- OnFlushCompleted
- OnCompactionCompleted

As @paolococchi requested in #4288 , and @ajkr agreed, we should also support `OnCompactionBegin`.

This PR is a try to implement the support of `OnCompactionBegin`.

Hope it is useful to you.

### CODE REVIEW

**Comment by ajkr** [comment]
nice!

**Comment by solicomo** [comment]
It seems that the test failure "DBTest::SoftLimit" is not related to this fix.
#4432 also encounters this failure.

So I just rebase and give it another try.

**Comment by solicomo** [comment]
Thanks @riversand963 for your review.

I am working on the unit test.

File: db/db_impl_compaction_flush.cc [hunk 1]
```diff
@@ -744,6 +744,69 @@ Status DBImpl::ContinueBackgroundWork() {
   return Status::OK();
 }
 
+void DBImpl::NotifyOnCompactionBegin(ColumnFamilyData* cfd,
+                                     Compaction *c, const Status &st,
+                                     const CompactionJobStats& job_stats,
+                                     int job_id) {
+#ifndef ROCKSDB_LITE
+  if (immutable_db_options_.listeners.size() == 0U) {
```
[the remark this answers is not in the document]
Comment (solicomo) [reply db/db_impl_compaction_flush.cc]: Done.

File: db/db_impl_compaction_flush.cc [hunk 2]
```diff
@@ -744,6 +744,69 @@ Status DBImpl::ContinueBackgroundWork() {
...
+  }
+  mutex_.AssertHeld();
+  if (shutting_down_.load(std::memory_order_acquire)) {
+    return;
+  }
+  Version* current = cfd->current();
+  current->Ref();
+  // release lock while notifying events
+  mutex_.Unlock();
+  TEST_SYNC_POINT("DBImpl::NotifyOnCompactionBegin::UnlockMutex");
```
[the remark this answers is not in the document]
Comment (solicomo) [reply db/db_impl_compaction_flush.cc]: Sorry for the delay of update. 
It took me a long time to figure out how is the unit test working.
I'm still not sure my understanding is correct.

Anyway, I update an existing unit test and use this sync point and confirm that `OnCompactionBegin` is working.

Please leave a comment if you think anything is incorrect or can be refined.
Thank you.

File: db/db_impl_compaction_flush.cc [hunk 3]
```diff
@@ -744,6 +744,69 @@ Status DBImpl::ContinueBackgroundWork() {
   return Status::OK();
 }
 
+void DBImpl::NotifyOnCompactionBegin(ColumnFamilyData* cfd,
```
[the remark this answers is not in the document]
Comment (solicomo) [reply db/db_impl_compaction_flush.cc]: To let the PR simple and clean, I'd like leave it as it is right now.
I will create another PR to refine the redundant code.
How do you think?

Thank you.

[the remark this answers is not in the document]
Comment (riversand963) [reply db/db_impl_compaction_flush.cc]: SG

**Comment by solicomo** [comment]
Thanks @riversand963 for your review and suggestions.

**Comment by yiwu-arbug** [comment]
@riversand963 Is this ready to land?

**Comment by riversand963** [comment]
@yiwu-arbug , I think so.

**Comment by riversand963** [comment]
Land initiated. @yiwu-arbug

**Pull request closed without merging**